src/ir/term/mod.rs

```diff
@@ -721,6 +723,104 @@ impl Array {
         self.check_idx(idx);
         self.map.get(idx).unwrap_or(&*self.default).clone()
     }
+
+    /// Iter
+    pub fn into_iter(&self) -> std::collections::btree_map::IntoIter<Value, Value> {
+        self.map.clone().into_iter()
+    }
+}
+
+/// Merge two Array Iterators
+pub struct ArrayMerge {
+    left: Peekable<Box<std::collections::btree_map::IntoIter<Value, Value>>>,
+    right: Peekable<Box<std::collections::btree_map::IntoIter<Value, Value>>>,
+    left_dfl: Value,
+    right_dfl: Value,
+}
+
+impl ArrayMerge {
+    /// Create a new [ArrayMerge] from two [Array]s
+    pub fn new(a: Array, b: Array) -> Self {
+        if a.size != b.size {
+            panic!("IR Arrays have different lengths: {}, {}", a.size, b.size);
+        }
+        if a.key_sort != b.key_sort {
+            panic!(
+                "IR Arrays have different key sorts: {}, {}",
+                a.key_sort, b.key_sort
+            );
+        }
+        if a.default.sort() != b.default.sort() {
+            panic!(
+                "IR Arrays default values have different key sorts: {}, {}",
+                a.default.sort(),
+                b.default.sort()
+            );
+        }
+
+        Self {
+            left: Box::new(a.into_iter()).peekable(),
+            right: Box::new(b.into_iter()).peekable(),
```

Comment:
I doubt you need `Box` here even in the current implementation. You would need a `Box` if you wanted to make this generic over any iterator (because in that case it would be a trait object, `Box<dyn ...>` or so). But here you're referring to a concrete type that's sized, so you definitely don't need to box it, and you don't need the extra level of indirection that comes with a heap-allocated object.

This will be true after converting to work with references and implementing `IntoIterator` as described above---Array.into_iter() returns a concrete type (`std::collection::btree_map::Iter<'a, Value, Value>`), and that's the type that this struct will hold.